client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/iteration/RemotePublisher.java

```diff
@@ -55,15 +56,15 @@ public RemotePublisher(OperationsFactory operationsFactory, String filterConvert
       if (segments == null) {
          if (segmentConsistentHash != null) {
             int maxSegment = segmentConsistentHash.getNumSegments();
-            this.segments = IntSets.mutableEmptySet(maxSegment);
+            this.segments = IntSets.concurrentSet(maxSegment);
```

Comment:
These two changes here look like the correct fix. 

Looking at the code, I am pretty sure the actual issue is that when `handleResponse` is called is a different thread and not synchronized with the publisher, thus causing updates to `segments` to not always be visible, which can cause duplicate values processed for a given set of segments.